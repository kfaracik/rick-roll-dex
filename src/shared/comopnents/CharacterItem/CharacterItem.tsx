import React from 'react';
import {GestureResponderEvent, Image, View} from 'react-native';
import {useAtom} from 'jotai';
import {styles} from './CharacterItem.styled';
import {CategoryValueText, Button, Card} from '..';
import {Character} from '../../api';
import {Colors} from '../../utils';
import {favoriteCharactersAtom} from '../../store/favoriteCharactersStore';

type CharacterItemProps = Pick<
  Character,
  'id' | 'name' | 'status' | 'species' | 'image'
> & {
  onPress: () => void;
};

export const CharacterItem = ({
  id,
  name,
  status,
  species,
  image,
  onPress,
}: CharacterItemProps) => {
  const [favorites, setFavorites] = useAtom(favoriteCharactersAtom);

  const liked = favorites.includes(id);

  const onLikePress = (event: GestureResponderEvent | undefined) => {
    event?.stopPropagation();
    setFavorites(prev =>
      liked ? prev.filter(favId => favId !== id) : [...prev, id],
    );
  };

  return (
    <Card onPress={onPress} contentStyle={styles.cardContent}>
      <View style={styles.textContainer}>
        <CategoryValueText category="NAME" value={name} />
        <CategoryValueText category="STATUS" value={status} />
        <CategoryValueText category="SPECIES" value={species} />
      </View>
      <View>
        <Button
          title={'Like'}
          onPress={onLikePress}
          mode="white"
          iconName={liked ? 'star' : 'staro'}
          iconColor={liked ? Colors.accent : Colors.primary}
          style={styles.likeButton}
        />
        <Image source={{uri: image}} style={styles.image} />
      </View>
    </Card>
  );
};
