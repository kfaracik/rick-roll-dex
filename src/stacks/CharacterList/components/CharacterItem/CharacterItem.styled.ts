import {StyleSheet} from 'react-native';
import {Colors} from '../../../../shared/utils';

export const styles = StyleSheet.create({
  card: {
    height: 224,
    borderRadius: 24,
    borderColor: Colors.primary,
    borderWidth: 1,
    overflow: 'hidden',
    marginHorizontal: 16,
    padding: 12,
  },
  image: {
    position: 'absolute',
    borderWidth: 1,
    borderColor: Colors.primary,
    borderRadius: 24,
    right: 0,
    minWidth: 200,
    minHeight: 200,
  },
  content: {
    height: '100%',
    padding: 0,
    margin: 0,
  },
  likeButton: {
    position: 'absolute',
    bottom: 8,
    right: 8,
    borderRadius: 24,
    backgroundColor: Colors.white,
    borderWidth: 1,
    borderColor: Colors.primary,
  },
});
