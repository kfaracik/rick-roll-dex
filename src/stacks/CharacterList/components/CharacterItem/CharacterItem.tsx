import React from 'react';
import {Alert, GestureResponderEvent, Image, View} from 'react-native';
import {styles} from './CharacterItem.styled';
import {CategoryValueText, Button, Card} from '../../../../shared/comopnents';
import {Character} from '../../../../shared/api';
import {Colors} from '../../../../shared/utils';

type CharacterItemProps = Pick<
  Character,
  'name' | 'status' | 'species' | 'image'
> & {
  liked: boolean;
  onPress: () => void;
};

export const CharacterItem = ({
  onPress,
  liked,
  name,
  status,
  species,
  image,
}: CharacterItemProps) => {
  const onLikePress = (event: GestureResponderEvent | undefined) => {
    event?.stopPropagation();
    Alert.alert('TODO implement like logic');
  };

  return (
    <Card onPress={onPress} contentStyle={styles.cardContent}>
      <View style={styles.textContainer}>
        <CategoryValueText category="NAME" value={name} />
        <CategoryValueText category="STATUS" value={status} />
        <CategoryValueText category="SPECIES" value={species} />
      </View>
      <View>
        <Button
          title="Like"
          onPress={onLikePress}
          mode="white"
          iconName={liked ? 'star' : 'staro'}
          iconColor={liked ? Colors.accent : Colors.primary}
          style={styles.likeButton}
        />
        <Image source={{uri: image}} style={styles.image} />
      </View>
    </Card>
  );
};
