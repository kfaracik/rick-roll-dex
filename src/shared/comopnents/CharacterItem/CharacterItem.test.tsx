import React from 'react';
import {render} from '@testing-library/react-native';
import {CharacterItem} from './CharacterItem';

describe('CharacterItem component', () => {
  it('should generate snapshot', () => {
    const {toJSON} = render(
      <CharacterItem
        image={'1234'}
        status={'Alive'}
        id={123}
        name={'naem'}
        species={'species'}
        onPress={jest.fn}
      />,
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
