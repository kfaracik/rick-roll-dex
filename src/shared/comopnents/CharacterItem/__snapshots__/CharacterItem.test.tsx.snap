// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterItem component should generate snapshot 1`] = `
<View
  collapsable={false}
  entering={
    FadeIn {
      "build": [Function],
      "durationV": 200,
      "randomizeDelay": false,
      "reduceMotionV": "system",
    }
  }
  exiting={
    FadeOut {
      "build": [Function],
      "durationV": 200,
      "randomizeDelay": false,
      "reduceMotionV": "system",
    }
  }
  jestAnimatedStyle={
    {
      "value": {},
    }
  }
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "rgba(247, 243, 249, 1)",
        "borderRadius": 24,
        "margin": 0,
        "marginHorizontal": 16,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.15,
        "shadowRadius": 3,
      }
    }
    testID="card-container-outer-layer"
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "rgba(247, 243, 249, 1)",
          "borderColor": "#224229",
          "borderRadius": 24,
          "borderWidth": 1,
          "boxShadow": [
            {
              "color": "#224229",
              "offsetX": 5,
              "offsetY": 5,
            },
          ],
          "elevation": 1,
          "flex": undefined,
          "minHeight": 224,
          "overflow": "hidden",
          "padding": 12,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 1,
            "width": 0,
          },
          "shadowOpacity": 0.3,
          "shadowRadius": 1,
        }
      }
      testID="card-container"
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View
          style={
            [
              {
                "flexShrink": 1,
              },
              undefined,
            ]
          }
          testID="card"
        >
          <View
            borderRadiusStyles={
              {
                "borderRadius": 24,
              }
            }
            style={
              [
                {
                  "paddingHorizontal": 16,
                },
                {
                  "paddingVertical": 16,
                },
                [
                  {
                    "alignItems": "flex-start",
                    "gap": 20,
                    "margin": 0,
                    "paddingHorizontal": 0,
                    "paddingVertical": 0,
                  },
                  {
                    "flexDirection": "row",
                  },
                ],
              ]
            }
          >
            <View
              style={
                {
                  "flex": 1,
                }
              }
            >
              <View
                style={
                  [
                    {
                      "padding": 8,
                    },
                    undefined,
                  ]
                }
              >
                <Text
                  style={
                    {
                      "color": "#59695C",
                      "fontFamily": "DMMono-Regular",
                      "fontSize": 12,
                    }
                  }
                >
                  NAME
                </Text>
                <Text
                  style={
                    [
                      {
                        "color": "#162C1B",
                        "fontFamily": "Inter_18pt-Regular",
                        "fontSize": 16,
                      },
                      undefined,
                    ]
                  }
                >
                  naem
                </Text>
              </View>
              <View
                style={
                  [
                    {
                      "padding": 8,
                    },
                    undefined,
                  ]
                }
              >
                <Text
                  style={
                    {
                      "color": "#59695C",
                      "fontFamily": "DMMono-Regular",
                      "fontSize": 12,
                    }
                  }
                >
                  STATUS
                </Text>
                <Text
                  style={
                    [
                      {
                        "color": "#162C1B",
                        "fontFamily": "Inter_18pt-Regular",
                        "fontSize": 16,
                      },
                      undefined,
                    ]
                  }
                >
                  Alive
                </Text>
              </View>
              <View
                style={
                  [
                    {
                      "padding": 8,
                    },
                    undefined,
                  ]
                }
              >
                <Text
                  style={
                    {
                      "color": "#59695C",
                      "fontFamily": "DMMono-Regular",
                      "fontSize": 12,
                    }
                  }
                >
                  SPECIES
                </Text>
                <Text
                  style={
                    [
                      {
                        "color": "#162C1B",
                        "fontFamily": "Inter_18pt-Regular",
                        "fontSize": 16,
                      },
                      undefined,
                    ]
                  }
                >
                  species
                </Text>
              </View>
            </View>
            <View>
              <View
                collapsable={false}
                style={
                  {
                    "alignSelf": "flex-start",
                    "backgroundColor": "#FFFFFF",
                    "borderRadius": 24,
                    "bottom": 8,
                    "height": 34,
                    "position": "absolute",
                    "right": 8,
                    "shadowColor": "#000",
                    "shadowOffset": {
                      "height": 0,
                      "width": 0,
                    },
                    "shadowOpacity": 0,
                    "shadowRadius": 0,
                  }
                }
                testID="button-container-outer-layer"
              >
                <View
                  collapsable={false}
                  style={
                    {
                      "backgroundColor": "#FFFFFF",
                      "borderColor": "#224229",
                      "borderRadius": 24,
                      "borderStyle": "solid",
                      "borderWidth": 1,
                      "flex": 1,
                      "minWidth": 64,
                      "rippleColor": "#DAE4DC",
                      "shadowColor": "#000",
                      "shadowOffset": {
                        "height": 0,
                        "width": 0,
                      },
                      "shadowOpacity": 0,
                      "shadowRadius": 0,
                      "zIndex": 1,
                    }
                  }
                  testID="button-container"
                >
                  <View
                    accessibilityRole="button"
                    accessibilityState={
                      {
                        "busy": undefined,
                        "checked": undefined,
                        "disabled": false,
                        "expanded": undefined,
                        "selected": undefined,
                      }
                    }
                    accessibilityValue={
                      {
                        "max": undefined,
                        "min": undefined,
                        "now": undefined,
                        "text": undefined,
                      }
                    }
                    accessible={true}
                    collapsable={false}
                    focusable={true}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      [
                        {
                          "overflow": "hidden",
                        },
                        {
                          "borderRadius": 24,
                        },
                      ]
                    }
                    testID="button"
                  >
                    <View
                      style={
                        [
                          {
                            "alignItems": "center",
                            "flexDirection": "row",
                            "justifyContent": "center",
                          },
                          {
                            "margin": -6,
                          },
                        ]
                      }
                    >
                      <View
                        style={
                          [
                            {
                              "marginLeft": 12,
                              "marginRight": -4,
                            },
                            {
                              "marginLeft": 16,
                              "marginRight": -16,
                            },
                            false,
                          ]
                        }
                        testID="button-icon-container"
                      >
                        <Text
                          allowFontScaling={false}
                          selectable={false}
                          style={
                            [
                              {
                                "color": "#224229",
                                "fontSize": 16,
                              },
                              undefined,
                              {
                                "fontFamily": "anticon",
                                "fontStyle": "normal",
                                "fontWeight": "normal",
                              },
                              {},
                            ]
                          }
                        >
                          
                        </Text>
                      </View>
                      <Text
                        numberOfLines={1}
                        selectable={false}
                        style={
                          [
                            {
                              "textAlign": "left",
                            },
                            {
                              "color": "rgba(28, 27, 31, 1)",
                              "writingDirection": "ltr",
                            },
                            [
                              [
                                {
                                  "marginHorizontal": 16,
                                  "marginVertical": 9,
                                  "textAlign": "center",
                                },
                                false,
                                {
                                  "marginHorizontal": 24,
                                  "marginVertical": 10,
                                },
                                undefined,
                                {
                                  "textTransform": "uppercase",
                                },
                                {
                                  "color": "rgba(255, 255, 255, 1)",
                                  "fontFamily": "System",
                                  "fontSize": 14,
                                  "fontWeight": "500",
                                  "letterSpacing": 0.1,
                                  "lineHeight": 20,
                                },
                                undefined,
                              ],
                              {
                                "color": "#224229",
                                "fontFamily": "DMMono-Regular",
                                "fontSize": 14,
                              },
                            ],
                          ]
                        }
                        testID="button-text"
                      >
                        <Text
                          style={
                            [
                              {
                                "textAlign": "left",
                              },
                              {
                                "color": "rgba(28, 27, 31, 1)",
                                "fontFamily": "System",
                                "fontWeight": "400",
                                "letterSpacing": 0,
                              },
                              {
                                "writingDirection": "ltr",
                              },
                              {
                                "color": "#224229",
                                "fontFamily": "DMMono-Regular",
                                "fontSize": 14,
                              },
                            ]
                          }
                        >
                          Like
                        </Text>
                      </Text>
                    </View>
                  </View>
                </View>
              </View>
              <Image
                source={
                  {
                    "uri": "1234",
                  }
                }
                style={
                  {
                    "borderColor": "#224229",
                    "borderRadius": 24,
                    "borderWidth": 1,
                    "height": 200,
                    "width": 200,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
